src/libCps/CpsRunEvm.cpp

```diff
@@ -0,0 +1,568 @@
+      mExecutor.PushRun(std::move(transferRun));
+    }
+  }
+
+  return {TxnStatus::NOT_PRESENT, true, {}};
+}
+
+CpsExecuteResult CpsRunEvm::ValidateCreateTrap(
+    const evm::TrapData_Create& createData, uint64_t remainingGas) {
+  if (mCpsContext.isStatic) {
+    return {TxnStatus::FAIL_CONTRACT_ACCOUNT_CREATION, false, {}};
+  }
+  const evm::Address& protoCaller = createData.caller();
+  const Address address = ProtoToAddress(protoCaller);
+
+  Address fromAddress;
+  const auto& scheme = createData.scheme();
+  if (scheme.has_legacy()) {
+    const evm::TrapData_Scheme_Legacy& legacy = scheme.legacy();
+    fromAddress = ProtoToAddress(legacy.caller());
+  } else if (scheme.has_create2()) {
+    const evm::TrapData_Scheme_Create2& create2 = scheme.create2();
+    fromAddress = ProtoToAddress(create2.caller());
+  } else if (scheme.has_fixed()) {
+    fromAddress = ProtoToAddress(mProtoArgs.address());
+  }
+
+  // Caller should be the same as the contract that triggered trap
+  const auto currentAddress = ProtoToAddress(mProtoArgs.address());
+  if (address != currentAddress || fromAddress != currentAddress) {
+    return {TxnStatus::INVALID_FROM_ACCOUNT, false, {}};
+  }
+
+  // Check Balance, Required Gas
+
+  const auto currentBalance = mAccountStore.GetBalanceForAccountAtomic(address);
+  const auto requestedValue = Amount::fromWei(ProtoToUint(createData.value()));
+  if (requestedValue > currentBalance) {
+    return {TxnStatus::INSUFFICIENT_BALANCE, false, {}};
+  }
+
+  const auto targetGas =
+      createData.target_gas() != std::numeric_limits<uint64_t>::max()
+          ? createData.target_gas()
+          : remainingGas;
+
+  auto baseFee = Eth::getGasUnitsForContractDeployment(
+      {}, DataConversion::StringToCharArray(createData.call_data()));
+
+  if (ProtoToUint(createData.value()) > 0) {
+    baseFee += MIN_ETH_GAS;
+  }
+  if (targetGas < baseFee || targetGas > remainingGas) {
+    return {TxnStatus::INSUFFICIENT_GAS_LIMIT, false, {}};
+  }
+
+  return {TxnStatus::NOT_PRESENT, true, {}};
+}
+
+void CpsRunEvm::HandleApply(const evm::EvmResult& result,
+                            TransactionReceipt& receipt) {
+  if (result.logs_size() > 0) {
+    Json::Value entry = Json::arrayValue;
+
+    for (const auto& log : result.logs()) {
+      Json::Value logJson;
+      logJson["address"] = "0x" + ProtoToAddress(log.address()).hex();
+      logJson["data"] = "0x" + boost::algorithm::hex(log.data());
+      Json::Value topics_array = Json::arrayValue;
+      for (const auto& topic : log.topics()) {
+        topics_array.append("0x" + ProtoToH256(topic).hex());
+      }
+      logJson["topics"] = topics_array;
+      entry.append(logJson);
+    }
+    receipt.AddJsonEntry(entry);
+  }
+
+  Address thisContractAddress = ProtoToAddress(mProtoArgs.address());
+  Address accountToRemove;
+  Address fundsRecipient;
+  Amount funds;
+
+  // parse the return values from the call to evm.
+  for (const auto& it : result.apply()) {
```

Comment:
I think it has more than one entry in case of selfdestruct (then we have a contract that should be removed and fund recipient). I left that in a loop otherwise we should specify another counter (not sure if result.apply() has known size).

Reply:
SG. document this please.